plugins/optimization-detective/class-od-element.php

```diff
@@ -0,0 +1,115 @@
+<?php
+/**
+ * Optimization Detective: OD_Element class
+ *
+ * @package optimization-detective
+ * @since n.e.x.t
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Data for a single element in a URL metric.
+ *
+ * @phpstan-import-type ElementData from OD_URL_Metric
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+class OD_Element implements ArrayAccess, JsonSerializable {
+
+	/**
+	 * Data.
+	 *
+	 * @since n.e.x.t
+	 * @var ElementData
+	 */
+	protected $data;
+
+	/**
+	 * URL metric that this element belongs to.
+	 *
+	 * @since n.e.x.t
+	 * @var OD_URL_Metric
+	 * @readonly
+	 */
+	public $url_metric;
+
+	/**
+	 * Constructor.
+	 *
+	 * @phpstan-param ElementData $data
+	 *
+	 * @param array<string, mixed> $data       Element data.
+	 * @param OD_URL_Metric        $url_metric URL metric.
+	 */
+	public function __construct( array $data, OD_URL_Metric $url_metric ) {
+		$this->data       = $data;
+		$this->url_metric = $url_metric;
+	}
+
+	/**
+	 * Checks whether an offset exists.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $offset Key.
+	 * @return bool Whether the offset exists.
+	 */
+	public function offsetExists( $offset ): bool {
+		return isset( $this->data[ $offset ] );
+	}
+
+	/**
+	 * Retrieves an offset.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $offset Key.
+	 * @return mixed Can return all value types.
+	 */
+	public function offsetGet( $offset ) { // phpcs:ignore SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint
```

Comment:
I implemented `ArrayAccess` here for back-compat with existing code accessing the offsets. But for the way we access data, perhaps we should have accessor `get` methods similar to what is done for `OD_URL_Metric`. So this would have:

* `is_lcp()`
* `is_lcp_candidate()`
* `get_xpath()`
* `get_intersection_ratio()`
* `get_intersection_rect()`
* `get_bounding_client_rect()`

As well as a `get()` method which would return arbitrary values, especially those which are from an extension-extended schema, like `$element->get( 'resizedBoundingClientRect )` in Embed Optimizer.

Reply:
Done in 3de4632f43ba79db5708a5127ad9c3d2d70b96b2